utils/npmutils.go

```diff
@@ -0,0 +1,303 @@
+		(*dependencies)[depId] = &entities.Dependency{Id: depId, Scopes: []string{scope}}
+	} else if !scopeAlreadyExists(scope, (*dependencies)[depId].Scopes) {
+		(*dependencies)[depId].Scopes = append((*dependencies)[depId].Scopes, scope)
+	}
+	(*dependencies)[depId].RequestedBy = append((*dependencies)[depId].RequestedBy, pathToRoot)
+}
+
+func scopeAlreadyExists(scope string, existingScopes []string) bool {
+	for _, existingScope := range existingScopes {
+		if existingScope == scope {
+			return true
+		}
+	}
+	return false
+}
+
+func runList(typeRestriction, executablePath, srcPath string, npmArgs []string, log Log) (stdResult, errResult []byte, err error) {
+	log.Debug("Running npm list command.")
+	cmdArgs := []string{"list"}
+	cmdArgs = append(cmdArgs, npmArgs...)
+
+	// These arguments must be added at the end of the command, to override their other values (if existed in nm.npmArgs)
+	cmdArgs = append(cmdArgs, "--json=true", "--all", "--"+typeRestriction)
+
+	command := exec.Command(executablePath, cmdArgs...)
+	command.Dir = srcPath
+	outBuffer := bytes.NewBuffer([]byte{})
+	command.Stdout = outBuffer
+	errBuffer := bytes.NewBuffer([]byte{})
+	command.Stderr = errBuffer
+	err = command.Run()
+	errResult = errBuffer.Bytes()
+	log.Debug("npm list error output is:\n" + string(errResult))
+	if err != nil {
+		if _, ok := err.(*exec.ExitError); ok {
+			err = errors.New(err.Error())
+		}
+		return
+	}
+	stdResult = outBuffer.Bytes()
+	log.Debug("npm list standard output is:\n" + string(stdResult))
+	return
+}
+
+func GetNpmVersionAndExecPath(log Log) (*version.Version, string, error) {
+	if log == nil {
+		log = &NullLog{}
+	}
+	npmExecPath, err := exec.LookPath("npm")
+	if err != nil {
+		return nil, "", err
+	}
+
+	if npmExecPath == "" {
+		return nil, "", errors.New("could not find the 'npm' executable in the system PATH")
+	}
+
+	log.Debug("Using npm executable:", npmExecPath)
+
+	npmVersion, err := getVersion(npmExecPath)
+	if err != nil {
+		return nil, "", err
+	}
+	log.Debug("Using npm version:", npmVersion)
+	return version.NewVersion(npmVersion), npmExecPath, nil
+}
+
+func getVersion(executablePath string) (string, error) {
+	command := exec.Command(executablePath, "-version")
+	buffer := bytes.NewBuffer([]byte{})
+	command.Stderr = buffer
+	command.Stdout = buffer
+	err := command.Run()
+	if _, ok := err.(*exec.ExitError); ok {
+		err = errors.New(err.Error())
+	}
+	return buffer.String(), err
+}
+
+type PackageInfo struct {
+	Name    string `json:"name,omitempty"`
+	Version string `json:"version,omitempty"`
+	Scope   string
+}
+
+func ReadPackageInfoFromPackageJson(packageJsonDirectory string, npmVersion *version.Version) (*PackageInfo, error) {
+	packageJson, err := ioutil.ReadFile(filepath.Join(packageJsonDirectory, "package.json"))
+	if err != nil {
+		return nil, err
+	}
+	return ReadPackageInfo(packageJson, npmVersion)
+}
+
+func ReadPackageInfo(data []byte, npmVersion *version.Version) (*PackageInfo, error) {
```

Comment:
Looks like thi sfunction can be made provate.

Reply:
It's also used jfrog-cli-core.